plugins/plugin_utils/_reboot.py

```diff
@@ -0,0 +1,374 @@
+in this collection. Right now it should only be used in ansible.windows as the
+interface is not final and could be subject to change.
+"""
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import datetime
+import random
+import time
+import traceback
+
+from ansible.errors import AnsibleConnectionFailure, AnsibleError
+from ansible.module_utils.common.text.converters import to_native, to_text
+from ansible.utils.display import Display
+
+from ansible.plugins.connection import ConnectionBase
+try:
+    from typing import (
+        Any,
+        Callable,
+        Dict,
+        Tuple,
+    )
+except ImportError:
+    # Satisfy Python 2 which doesn't have typing.
+    Any = Callable = Dict = Tuple = None
+
+
+# This is not ideal but the psrp connection plugin doesn't catch all these exceptions as an AnsibleConnectionFailure.
+# Until we can guarantee we are using a version of psrp that handles all this we try to handle those issues.
+try:
+    from requests.exceptions import (
+        ConnectionError as RequestsConnectionError,
+        Timeout as RequestsTimeout,
+    )
+except ImportError:
+    RequestsConnectionError = RequestsTimeout = None
+
+
+_DEFAULT_BOOT_TIME_COMMAND = "(Get-CimInstance -ClassName Win32_OperatingSystem -Property LastBootUpTime)" \
+                             ".LastBootUpTime.ToFileTime()"
+display = Display()
+
+
+class _ReturnResultException(Exception):
+
+    def __init__(self, msg, **result):
+        super(_ReturnResultException, self).__init__(msg)
+        self.result = result
+
+
+def reboot_action(
+        task_action,
+        connection,
+        boot_time_command=_DEFAULT_BOOT_TIME_COMMAND,
+        connection_timeout=5,
+        msg='Reboot initiated by Ansible',
+        post_reboot_delay=0,
+        pre_reboot_delay=2,
+        reboot_timeout=600,
+        test_command='whoami',
+):  # type: (str, ConnectionBase, str, int, str, int, int, int, str) -> Dict
+    """Reboot a Windows Host.
+
+    Used by action plugins in ansible.windows to reboot a Windows host. It
+    takes in the action plugin so it can run the commands on the targeted host
+    and monitor the reboot process. The return dict will have the following
+    keys set:
+
+        changed: Whether a change occurred (reboot was done)
+        elapsed: Seconds elapsed between the reboot and it coming back online
+        failed: Whether a failure occurred
+        rebooted: Whether the host was rebooted
+
+    When failed=True there may be more keys to give some information around
+    the failure like msg, exception. There are other keys that might be
+    returned as well but they are dependent on the failure that occurred.
+
+    Verbosity levels used:
+        2: Message when each reboot step is completed
+        3: Connection plugin operations and their results
+        5: Raw commands run and the results of those commands
+        Debug:
```

Comment:
love all this documentation, but this is especially nice to be able to find without digging through each call, thanks!